pkg/admissioncontroller/webhooks/auth/seed/graph/eventhandler_bastion.go

```diff
@@ -0,0 +1,93 @@
+// Copyright (c) 2021 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package graph
+
+import (
+	"time"
+
+	gardencorev1alpha1 "github.com/gardener/gardener/pkg/apis/core/v1alpha1"
+
+	toolscache "k8s.io/client-go/tools/cache"
+	"sigs.k8s.io/controller-runtime/pkg/cache"
+)
+
+func (g *graph) setupBastionWatch(informer cache.Informer) {
+	informer.AddEventHandler(toolscache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			bastion, ok := obj.(*gardencorev1alpha1.Bastion)
+			if !ok {
+				return
+			}
+			g.handleBastionCreateOrUpdate(bastion)
+		},
+
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			oldBastion, ok := oldObj.(*gardencorev1alpha1.Bastion)
+			if !ok {
+				return
+			}
+
+			newBastion, ok := newObj.(*gardencorev1alpha1.Bastion)
+			if !ok {
+				return
+			}
+
+			if oldBastion.Status.SeedName != newBastion.Status.SeedName {
+				g.handleBastionCreateOrUpdate(newBastion)
+			}
+		},
+
+		DeleteFunc: func(obj interface{}) {
+			if tombstone, ok := obj.(toolscache.DeletedFinalStateUnknown); ok {
+				obj = tombstone.Obj
+			}
+			bastion, ok := obj.(*gardencorev1alpha1.Bastion)
+			if !ok {
+				return
+			}
+			g.handleBastionDelete(bastion)
+		},
+	})
+}
+
+func (g *graph) handleBastionCreateOrUpdate(bastion *gardencorev1alpha1.Bastion) {
+	start := time.Now()
+	defer func() {
+		metricUpdateDuration.WithLabelValues("Bastion", "CreateOrUpdate").Observe(time.Since(start).Seconds())
+	}()
+	g.lock.Lock()
+	defer g.lock.Unlock()
+
+	g.deleteAllOutgoingEdges(VertexTypeBastion, bastion.Namespace, bastion.Name, VertexTypeSeed)
```

Comment:
According to https://github.com/gardener/gardener/blob/master/docs/deployment/gardenlet_api_access.md#pitfalls, I guess this line is not needed.

Reply:
Without this line, the graph unit tests didn't pass for some reason. I was also expecting that removing the vertex would remove all edges to it, but that was not the case. I can remove it to demo the error, if you like :)

Reply:
Removing the vertex calls https://github.com/gardener/gardener/blob/dd4154eefb3f0d842f01b49594e9008b55c1974b/pkg/admissioncontroller/webhooks/auth/seed/graph/graph.go#L189 which states:

> RemoveNode removes the node with the given ID from the graph, as well as any edges attached to it. If the node is not in the graph it is a no-op.

So, I'm not sure why the tests didn't pass anymore, maybe you can check again and if you still don't succeed, I'll try to also take a detailed look beginning of next week.

Reply:
You were right, the line was not needed.